### ISSUE TITLE
Possibility to set proxy start on jupyter machine 

### ISSUE THREAD

**Issue by BobCashStory** [jupyterhub/jupyter-server-proxy#218]
### Proposed change

I would like to start a process when user start they machine .
But for now the process start only when the user click on the launcher button, so can we have an `autosatart` option to start on machine boot ?

Or can the proxy allow to connect to already running app with the launcher entry present ?

**Comment by BobCashStory** [comment]
For people who want the same, i did a hack :

install port forwarding module :
`apt-get install redir`
add to your jupyter_notebook_config.py
```python
import os
os.system('python -m naas.runner -d') # Your process in detached mode

c.ServerProxy.servers = {
    'naas': {
        'launcher_entry': {
            'enabled': True,
            'icon_path': '/etc/jupyter/naas_logo.svg',
            'title': 'Naas manager',
        },
        'new_browser_tab': False,
        'timeout': 30,
        'command': ["redir", "--lport={port}", "--cport=5000"], # the redirect command to it !
    }
}
```

**Comment by ryanlovett** [comment]
Hi @BobCashStory , I think starting at boot is out of scope for jupyter-server-proxy, but you can always do this via the init system, e.g. a systemd unit file.

jupyter-server-proxy can already view an existing service however. For example you can visit /proxy/5000/. We don't have an elegant way to do this via a launcher however since we do require the `command` in the config.

**Comment by psychemedia** [comment]
If you are working in a BinderHub environment, you can add startup items to the `.binder/start` file:

#!/bin/bash

nohup mycommand subcommand &

exec "$@"

**Comment by consideRatio** [comment]
## A user experience (UX) motivating autostarting processes

I'm installing [syncthing](https://anaconda.org/conda-forge/syncthing) which is like dropbox - syncing files/directories. A service like syncthing need to be started to do it's configured directory syncing, but it also needs to be accessed by users that initially configures what directories are to be synced etc.

[jupyter-syncthing-proxy](https://github.com/yuvipanda/jupyter-syncthing-proxy/blob/main/jupyter_syncthing_proxy/__init__.py) helps expose the syncthing UI to configure it, but also starts syncthing by doing so.

Currently, the jupyterhub where i install this just adds `syncthing` to be installed via conda-forge, and `jupyter-syncthing-proxy` via PyPI. I think another step would be relevant though - to ensure syncthing gets started so previously configured directory syncing can resume on startup without needing to press the launcher button.

## Feature description

The feature as I think of it, would be to allow `jupyter-server-proxy` to automatically start a command ahead of the first request after having loaded its configuration. So I distinguish boot of the computer/container from having loaded of `jupyter-server-proxy`. I presume jupyter-server-proxy would start as part of a jupyter_server starts.

A benefit of not working around jupyter-server-proxy's inability to autostart on load would be that you wouldn't start two separate processes - one that jupyter-server-proxy wasn't aware of and one that it was aware of.

---

@ryanlovett what do you think about autostart as a feature on jupyter-server-proxy initialization rather than system boot?

**Comment by psychemedia** [comment]
Another proxied service that might benefit from automated start-up is a notebook search engine that can be started when the server starts so that it can index notebooks as they are edited. I had an early POC [here](https://github.com/ouseful-testing/nbsearch/) that IIRC had a horrible start up routine that blocked notebook server startup IIRC as it tried to index things before starting up the Jupyter server.

**Comment by ryanlovett** [comment]
@consideRatio That makes logical sense to me -- I have no objections. `jupyter-server-proxy` is both a web proxy and a process manager (via @yuvipanda's [simpervisor](https://github.com/jupyterhub/simpervisor)), which sounds like an odd pairing, but it works since we're not trying to do much of the latter.

As an alternative, what if there was a jupyter-server extension that could run configurable hooks at server startup? In this case, it would issue a GET onto whatever proxied endpoint is desired, causing the corresponding process to start. This is decoupled from the `jupyter-server-proxy` configuration, and maybe that's not desired or too messy. Just food for thought.

**Comment by manics** [comment]
My first thought is it sounds like a good idea, but I also think we should first scope out what it does, and what will/won't be supported in future to prevent maintenance becoming a burden.

- Will the autostart be for web-services only (IIRC jupyter-server-proxy checks the process is alive by making a network request) or will non-network services also be supported (maybe you want to run some background tasks that don't require interaction?)
- Is this something that should be in a separate component that jupyter-server-proxy builds on top of, e.g. so people can take advantage of process management without exposing any proxying URLs?

**Comment by luanabeckerdaluz** [comment]
I have the same question as @ghansham. I need to start a jupyter-server-proxy from inside notebook. 

I could start a flask app inside the notebook, but I need to invoke `window.open` from javascript and user need to reach `<jupyterhub_host>/user/<username>/proxy/{port}` and not `http://localhost/{port}`. That is why I want to start and stop a jupyter-server-proxy server. Is there any way to do this?

Cheers,
Luana
